### PR TITLE
Bluetooth: Controller: Fix assert on aux LLL scheduled chain reception

### PR DESCRIPTION
**Pull request by cvinayak**
Fix asserted in ULL due to incorrect resumption of scan
window when auxiliary channel chain PDU is LLL scheduled by
a ULL scheduled auxiliary channel PDU reception.

Relates to #38146.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

**Review by carlescufi — APPROVED** [review]

**Comment by cfriedt** [comment]
@thoh-ot - can you please have a look?

**Review by ppryga-nordic — CHANGES_REQUESTED** [review]
Could you provide a description, in a step-by-step way, of the failing scenario?
Then put a description how is it solved by the commit.

The single sentence description is a bit hard to follow. Taking into account switching in execution path between ULL and LLL to get clear view of the scenario is hard.

**Comment by cvinayak** [comment]
> Could you provide a description, in a step-by-step way, of the failing scenario? Then put a description how is it solved by the commit.
> 

@ppryga, I had the below in commit message:
> ... incorrect resumption of scan
window when auxiliary channel chain PDU is LLL scheduled by
a ULL scheduled auxiliary channel PDU reception

Is it ok to add the following:
_The issue is solved by having `is_chain_sched` flag in the auxiliary channel scan context and using the already present `is_aux_sched` in the primary channel scan to differentiate if the auxiliary PDU Rx ISR is to return back to primary channel scan window or to close the auxiliary chain PDU reception radio event._

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c [hunk 1]
```diff
@@ -586,33 +586,36 @@ void lll_scan_aux_isr_aux_setup(void *param)
...
+	/* scan context has used LLL scheduling for aux reception */
+	if (lll->is_aux_sched) {
+		isr_rx(lll, NULL, phy_aux);
 	} else {
-		phy_aux = node_rx->hdr.rx_ftr.aux_phy;
+		/* `lll->lll_aux` would be allocated in ULL for LLL scheduled
+		 * auxiliary PDU reception by scan context and for case
+		 * where LLL scheduled chain PDU reception by aux context, it
+		 * is assigned with the current aux context's LLL context.
+		 */
```
Comment (ppryga) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:616]:
```suggestion
		/* `lll->lll_aux` is allocated by ULL when auxiliary scan is started.
		 * It is used later by LLL in auxiliary scan context for scheduling
		 * reception of chained PDUs.
		 */
```

Correct me if I'm wrong, it may not happen that this path is executed with `lll->lll_aux == NULL`. It may be executed only in case the `isr_rx_lll_schedule` is set for reception of chained auxiliary PDU.

**Review by Thalley — APPROVED** [review]

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c [hunk 2]
```diff
@@ -1051,18 +1062,33 @@ static int isr_rx_pdu(struct lll_scan *lll, struct lll_scan_aux *lll_aux,
 	/* Passive scanner or scan responses */
 #if defined(CONFIG_BT_CENTRAL)
 	} else if (!lll->conn &&
-		   lll_scan_ext_tgta_check(lll, false, false, pdu, rl_idx)) {
+		   ((lll_aux && lll_aux->is_chain_sched) ||
+		    (lll->lll_aux && lll->lll_aux->is_chain_sched) ||
```
Comment (ppryga) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:1066]: Under what conditions is it possible to get here when only one of these conditions is not true `lll_aux && lll_aux->is_chain_sched`  or `lll->lll_aux && lll->lll_aux->is_chain_sched`?

**Review by ppryga-nordic — COMMENTED** [review]
One question and one comment. 
Sorry for nit-picking but getting a picture of the whole thing is a bit difficult.

Referring to updated comment in the commit. Thank you, that one is better.

**Review by Thalley — APPROVED** [review]

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c (hunk 1 above)
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:616]: There are two possible scenarios here:
1. The function is executed in a path that starts in LLL scan even context. In this situation `lll->is_aux_sched == true` and `else` statement does not fire. It does not matter if it is first or second received AUX_ADV_IND.
2. The function is executed in a path that starts in ULL auxiliary scan context. In this situation `else` statement fires and `lll->lll_aux` is already allocated by ULL. This is a reception of first chained PDU where TMASF is shorter than overhead.
3. The function is executed in a path that starts in LLL auxiliary scan context. In this situation `else` statement fires and `lll->lll_aux` is already allocated because it is a reception of consecutive chained PDU where TMASF is shorter than overhead. From brother perspective it may be seen as scenario that is following the one from point 2.

So the unwanted AUX_ADV_IND reception scheduled from LLL primary scan context fits into scenario 1.
Correct me if I'm wrong in my understanding?

Anyway, my questions clearly show that reasoning about code that handles extended and periodic advertising is not walk in the park. It includes multiple possible execution paths that are selected by number of closely named variables and executed by a number of functions that are reused in multiple paths. I would be more than happy If that is documented somewhere. It would help anyone who has to change/fix something in this part of controller code. 
And this comment actually could point to appropriate place in that documentation.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:616]: > 1. The function is executed in a path that starts in LLL scan even context. In this situation lll->is_aux_sched == true and else statement does not fire. It does not matter if it is first or second received AUX_ADV_IND.

Correction:
The path is `ULL scheduled primary scan context` --> `lll->is_aux_sched == true` --> `isr_rx_lll_schedule` --> `isr_rx(lll, NULL, phy_aux);` 

> 2. The function is executed in a path that starts in ULL auxiliary scan context. In this situation else statement fires and lll->lll_aux is already allocated by ULL. This is a reception of first chained PDU where TMASF is shorter than overhead.

The path is `ULL scheduled auxiliary scan context` --> `lll->is_aux_sched == false` -->`lll_aux->is_chain_sched == true` --> `isr_rx_lll_schedule`  --> `isr_rx(lll, lll->lll_aux, phy_aux);`

> 3. The function is executed in a path that starts in LLL auxiliary scan context. In this situation else statement fires and lll->lll_aux is already allocated because it is a reception of consecutive chained PDU where TMASF is shorter than overhead. From brother perspective it may be seen as scenario that is following the one from point 2.

This path starts after scenario 1 or 2, `LLL scheduled auxiliary scan context` --> `lll->is_aux_sched == false` AND `lll_aux->is_chain_sched == true` --> `isr_rx_lll_schedule`  --> `isr_rx(lll, lll->lll_aux, phy_aux);`

> So the unwanted AUX_ADV_IND reception scheduled from LLL primary scan context fits into scenario 1.
Correct me if I'm wrong in my understanding?

Yes, you are correct, in the scenario 1, `ULL scheduled primary scan context` --> `lll->is_aux_sched == true` --> `isr_rx_lll_schedule` --> `isr_rx(lll, NULL, phy_aux);` --> `lll->lll_aux == NULL` --> `return -ECHILD;` --> `radio_isr_set(lll_scan_isr_resume, lll);`

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c (hunk 2 above)
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:1066]: One more question here. `lll_aux == NULL` and lll->lll_aux == NULL` may happen only in case of execution path started by LLL primary scan context and a function `ull_scan_aux_setup()` was not called by ULL in the meantime (between LLL primary scan context and call to isr_rx_pdu) or the function was not able to allocate lll_scan_aux . It that correct?

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c (hunk 1 above)
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:616]: >This path starts after scenario 1 or 2, LLL scheduled auxiliary scan context --> lll->is_aux_sched == false AND lll_aux->is_chain_sched == true --> isr_rx_lll_schedule --> isr_rx(lll, lll->lll_aux, phy_aux);

How is it possible to start this path after scenario 1? The `lll->is_aux_sched` should be set to true until next primary scan prepare_cb.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:616]: You are correct, for scenario1 the continuation is, `LLL scheduled auxiliary scan context` --> `lll->is_aux_sched == true` AND `lll_aux->is_chain_sched == true` --> isr_rx_lll_schedule --> isr_rx(lll, NULL, phy_aux); and `lll->lll_aux` is valid.

Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:616]: That means the comment in the `else` statement is not correct. 
There is no possibility to execute `else` statement where `lll_aux == NULL`.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:616]: > That means the comment in the else statement is not correct.

The comment in the else is correct. Let me copy all the scenarios again, so that you can review and understand them:

1. `ULL scheduled primary scan context` --> `lll->is_aux_sched == true` --> `isr_rx_lll_schedule` --> `isr_rx(lll, NULL, phy_aux);` , and `lll->lll_aux` can be NULL if ULL could not allocate new aux context.

2. `ULL scheduled auxiliary scan context` --> `lll->is_aux_sched == false` -->`lll_aux->is_chain_sched == true` --> `isr_rx_lll_schedule`  --> `isr_rx(lll, lll->lll_aux, phy_aux);`

3. `LLL scheduled auxiliary scan context` --> `lll->is_aux_sched == false` AND `lll_aux->is_chain_sched == true` --> `isr_rx_lll_schedule`  --> `isr_rx(lll, lll->lll_aux, phy_aux);` and `lll->lll_aux` is always valid.  (`lll->is_aux_sched = false` for scenario 1 when primary scan starts while chains are being received).

4. `LLL scheduled auxiliary scan context` --> `lll->is_aux_sched == true` AND `lll_aux->is_chain_sched == true` --> isr_rx_lll_schedule --> isr_rx(lll, NULL, phy_aux); and `lll->lll_aux` is always valid.

These are the cases tested on target when scanning 2 different advertiser implementations, but there may be other combinations possible, but will handle it with more testing.

Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:616]: > 2. ULL scheduled auxiliary scan context --> lll->is_aux_sched == false -->lll_aux->is_chain_sched == true --> isr_rx_lll_schedule --> isr_rx(lll, lll->lll_aux, phy_aux);

Well, in my opinion this is not possible. I couldn't find a path that executes `isr_rx_lll_schedule` from ULL auxiliary context. It can be scheduled for execution from LLL auxiliary PDU contex as a reception for AUX_CHAIN_IND PDU first extended adv. PDU. But that is scenario 3...

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:616]: >  This is a reception of first chained PDU where TMASF is shorter than overhead.

You are concentrating on your above statement, but please consider the reception of AUX_ADV_IND that is placed greater than the threshold from the ADV_EXT_IND. Scenario 1, 2 and 3 are in sequence for the reception of ADV_EXT_IND, AUX_ADV_IND and AUX_CHAIN_IND respectively.

I am attaching an updated flowchart to reflect why the code comment is correct. 
[Extended Scanning.pdf](https://github.com/zephyrproject-rtos/zephyr/files/7264401/Extended.Scanning.pdf)

Also, please refer to original top-half of the comment here: https://github.com/zephyrproject-rtos/zephyr/pull/38490/files#diff-5ec58b9af07d51405994b09e2b6f185d144196c64a1f8728a791f3b2598892f5R402-R405

Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:616]: Me and @cvinayak had a private talk about the topic. @cvinayak is correct about flow. I was missing one thing in it. 
The comment in `else` statement is correct.

@cvinayak thank you for long and full of details discussion.

**Review by ppryga-nordic — APPROVED** [review]
